### PR TITLE
node-termination-by-default-from-15-0-0

### PR DESCRIPTION
**Pull request by calvix**
towards https://github.com/giantswarm/giantswarm/issues/16843

add validating webhook to allow only values `true` or `false` for the new annotation

add mutating webhook to migrate from old annotation (if present)
as in the previous alpha version, the annotation could be anything to enable the feature now we set the value to `true` if there is anything except `false`

code could be removed once all clusters are on 15.0.0

### CODE REVIEW

**Comment by calvix** [comment]
tested

File: pkg/aws/awscluster/mutate_awscluster.go [hunk 1]
```diff
@@ -431,6 +443,42 @@ func (m *Mutator) MutateRegion(awsCluster infrastructurev1alpha2.AWSCluster) ([]
...
+func (m *Mutator) MutateAnnotationNodeTerminateUnhealthy(awsCluster infrastructurev1alpha2.AWSCluster) ([]mutator.PatchOperation, error) {
+	var result []mutator.PatchOperation
+
+	release, err := semver.New(key.Release(&awsCluster))
+	if err != nil {
+		return nil, microerror.Maskf(notAllowedError, fmt.Sprintf("AWSCluster release has invalid value '%s'.", key.Release(&awsCluster)))
+	}
+
+	// new annotation is available from release >= 15.x.x
+	release15 := semver.MustParse("14.99.99")
```
Comment (calvix) [inline pkg/aws/awscluster/mutate_awscluster.go:457]: I had do this trick as when testing `15.0.0` was greater than `15.0.0-calvix` and it kinda breaks my testing, this should be safe

**Review by anvddriesch — COMMENTED** [review]
I think it looks good so far. However, I suggest adding a unit test to ensure the update works correctly.
Also, do we want this annotation to be on all the clusters from version 15 onwards? If so, where is it set for clusters which did not have the alpha annotation previously and newly created clusters?

**Comment by calvix** [comment]
>  think it looks good so far. However, I suggest adding a unit test to ensure the update works correctly.
Also, do we want this annotation to be on all the clusters from version 15 onwards? If so, where is it set for clusters which did not have the alpha annotation previously and newly created clusters?


I don't see the reason why we should add it to all clusters, if the annotations is not present aws-operator will assume the feature is enabled and act on it. It is just for the migration path.

I will work on the unit test

**Comment by anvddriesch** [comment]
Ah okay, thank you for the explanation, I wasn't sure what aws-operator expects. In that case this should cover everything.

**Review by njuettner — COMMENTED** [review]
some minor nits

File: pkg/aws/awscluster/validate_awscluster.go [hunk 2]
```diff
@@ -137,6 +141,22 @@ func (v *Validator) AWSClusterAnnotationPauseTimeIsValid(awsCluster infrastructu
 	return nil
 }
 
+func (v *Validator) AWSClusterAnnotationNodeTerminateUnhealthy(awsCluster infrastructurev1alpha2.AWSCluster) error {
```
[the remark this answers is not in the document]
Comment (calvix) [reply pkg/aws/awscluster/validate_awscluster.go:144]: we are currently have no unit tests for validation for awscluster and I am not certain how to build a new one without example :crying_cat_face:

File: pkg/aws/awscluster/mutate_awscluster.go [hunk 3]
```diff
@@ -431,6 +443,42 @@ func (m *Mutator) MutateRegion(awsCluster infrastructurev1alpha2.AWSCluster) ([]
 	return result, nil
 }
 
+//MutateAnnotationNodeTerminateUnhealthy migrate NodeTerminateUnhealthy annotations from alpha to stable in case it is configured.
+// this migration code can be removed once all AWS clusters are on release 15.0.0 or newer
+func (m *Mutator) MutateAnnotationNodeTerminateUnhealthy(awsCluster infrastructurev1alpha2.AWSCluster) ([]mutator.PatchOperation, error) {
```
[the remark this answers is not in the document]
Comment (calvix) [reply pkg/aws/awscluster/mutate_awscluster.go:448]: done

File: pkg/aws/awscluster/validate_awscluster.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (njuettner) [reply pkg/aws/awscluster/validate_awscluster.go:144]: Ok no problem. We can create a follow-up issue to add some tests.

[the remark this answers is not in the document]
Comment (calvix) [reply pkg/aws/awscluster/validate_awscluster.go:144]: https://github.com/giantswarm/giantswarm/issues/17396
created

File: pkg/aws/awscluster/mutate_awscluster.go [hunk 4]
```diff
@@ -431,6 +443,42 @@ func (m *Mutator) MutateRegion(awsCluster infrastructurev1alpha2.AWSCluster) ([]
 	return result, nil
 }
 
+//MutateAnnotationNodeTerminateUnhealthy migrate NodeTerminateUnhealthy annotations from alpha to stable in case it is configured.
+// this migration code can be removed once all AWS clusters are on release 15.0.0 or newer
```
[the remark this answers is not in the document]
Comment (calvix) [reply pkg/aws/awscluster/mutate_awscluster.go:448]: https://github.com/giantswarm/giantswarm/issues/17395

done

**Review by njuettner — APPROVED** [review]
lgtm

**Pull request merged**